cmd/containerd-shim-runhcs-v1/task_hcs.go

```diff
@@ -709,3 +696,119 @@ func (ht *hcsTask) Stats(ctx context.Context) (*stats.Statistics, error) {
 	}
 	return s, nil
 }
+
+func (ht *hcsTask) Update(ctx context.Context, req *task.UpdateTaskRequest) error {
+	resources, err := typeurl.UnmarshalAny(req.Resources)
+	if err != nil {
+		return errors.Wrapf(err, "failed to unmarshal resources for container %s update request", req.ID)
+	}
+
+	if err := verifyTaskUpdateResourcesType(resources); err != nil {
+		return err
+	}
+
+	if ht.ownsHost && ht.host != nil {
+		if err := ht.host.UpdateConstraints(ctx, resources, req.Annotations); err != nil {
+			return err
+		}
+	} else {
+		if err := ht.updateTaskContainerResources(ctx, resources, req.Annotations); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ht *hcsTask) updateTaskContainerResources(ctx context.Context, data interface{}, annotations map[string]string) error {
+	if ht.isWCOW {
+		return ht.updateWCOWResources(ctx, data, annotations)
+	}
+
+	return ht.updateLCOWResources(ctx, data, annotations)
+}
+
+func (ht *hcsTask) updateWCOWContainerCPU(ctx context.Context, cpu *specs.WindowsCPUResources) error {
+	// Add utility tool to container host if not present
+	hostPath, guestPath := getToolPaths()
+	shareReq := &shimdiag.ShareRequest{
+		HostPath: hostPath,
+		UvmPath:  guestPath,
+		ReadOnly: true,
+	}
+	err := ht.Share(ctx, shareReq)
+	if err == errTaskNotIsolated {
+		// process isolated task, just use host path to tool
+		guestPath = hostPath
+	} else if err != nil {
+		return err
+	}
+
+	args := createJobObjectsUtilArgs(ctx, guestPath, cpu)
+	req := &shimdiag.ExecProcessRequest{
+		Args: args,
+	}
+	if exitCode, err := ht.ExecInHost(ctx, req); err != nil {
+		return errors.Wrapf(err, "failed to exec command %v in host: %d", args, exitCode)
+	}
+
+	return nil
+}
+
+func isValidWindowsCPUResources(c *specs.WindowsCPUResources) bool {
+	return (c.Count != nil && (c.Shares == nil && c.Maximum == nil)) ||
+		(c.Shares != nil && (c.Count == nil && c.Maximum == nil)) ||
+		(c.Maximum != nil && (c.Count == nil && c.Shares == nil))
+}
+
+func (ht *hcsTask) updateWCOWResources(ctx context.Context, data interface{}, annotations map[string]string) error {
```

Comment:
I feel like we should unmarshal (or try to 😆) into *WindowsResources before this. I feel like if we're calling `updateWCOWResources` we've already hit some logic to end up here so by the time we get inside we should have the type ready to go. 

e.g. unmarshal here https://github.com/microsoft/hcsshim/pull/931/files#diff-04078e400c5f445fb34c4927b97930a0d1dbd00005819d7b7a77022d092f365cR723

Let me know what you think!

Reply:
us hitting logic that we're running a Windows Container doesn't guarantee that the sender actually sent a *WindowsResources struct. And since we need to check that case for both wcow and lcow, I just put it in their respective functions.

Reply:
No I know, was just saying maybe the caller should do the cast before calling this instead. To me it makes sense to have `updateWCOWResources` take in `*WindowsResources` (and same for lcow with its type).

Reply:
I wanted to make it open ended on the linux resources side specifically in case we ever need to pass both linux resources and windows resources to accomplish something.